### ISSUE TITLE
Allow explicitly building without mpi4py?

### ISSUE THREAD

**Issue by AlexanderRichert-NOAA** [Unidata/netcdf4-python#1389]
Version: 1.5.8
System: A Cray with SLES 15.3, but I expect this problem will occur on any system
Description of problem: The netcdf4-python installer auto-detects mpi4py by trying to load the module, and will build with it if available. This is a problem for us because we are trying to install through Spack, which requires having some Python version loaded, and our Python installation happens to include mpi4py _but_ we do not want netcdf4-python to use it (partly because we don't need it, and partly because if we wanted it, we would want it to point to an mpi4py copy inside our Spack-based installation, not the Python installation that's only used at build time). We are working on finding a way to modify our $PYTHONPATH to work around this issue, but would it be possible to have with/without mpi4py as an explicitly configurable build option as opposed to the current auto-detection scheme?

**Comment by jswhit** [comment]
Actually, setup.py looks for #define NC_HAS_PARALLEL in netcdf_meta.h to determine whether to add parrallel support to the build.  If it finds it, it then imports mpi4py to find out where the mpi4py header files are.  So, if you configure the spack-stack netcdf C library to not have parallel support then netcdf4-python won't have it either.

**Comment by jswhit** [comment]
also, the mpi4py that netcdf4-python uses at runtime doesn't need to be the same as the one it uses at build time.  The mpi4py module itself is not imported by netcdf4-python - it is not needed unless you're using parallel io.

**Comment by AlexanderRichert-NOAA** [comment]
Thanks for the info. I think what makes our case a bit unusual is that we are compiling NetCDF with MPI support, but the Python module without it. It sounds like maybe the Spack recipe should automatically determine MPI support based on whether NetCDF-C was built with it rather than having it as an independently configurable option.

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA FWIW, I ended up applying this patch to force building without parallel features, because I ran into this problem: https://github.com/nanoporetech/ont_fast5_api/issues/21 (with py-netcdf4/1.7.1-post2 and mpi4py 4.0.1). I think we need to apply the patch when ~mpi is set for py-netcdf4 (and we need to change our variant request in common packages.yaml).

@jswhit auto-detecting and guessing what user wants is more often than not not the best idea. There should be an explicit flag for py-netcdf4 that the user can set. If they want parallel capabilities, then check that everything is built with those options. Otherwise, don't care and build w/o parallel features.

**Comment by climbfuji** [comment]
@jswhit @AlexanderRichert-NOAA I wanted to check if there is any update on this issue. We still need to apply the patch if we want to build without MPI (and we only have a patch for version 1.7.2 in Spack). I am wondering if the original issue still exists (auto-detecting `mpi4py` in the latest develop code)?

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA Re-reading this a few times ... I think this problem is resolved as long as we *want* to build with MPI support enabled, because we build Python with Spack and the Spack-built Python does have `mpi4py` in the build environment for `py-netcdf4`.

**Comment by AlexanderRichert-NOAA** [comment]
As far as I'm aware our needs on this as far as operations haven't changed, i.e., we want the option of building without MPI but ideally without needing a separate NetCDF C spec. I agree that auto-guessing without the option to configure is not an ideal approach, but of course that's ultimately the developers' call. From my perspective, the Spack patch based on what Jeff provided is sufficient to work around the issue.

**Comment by jswhit** [comment]
I could have setup.py check for an envar (DISABLE_PARALLEL_SUPPORT) to over-ride the autodetection.  Would that work?

**Comment by AlexanderRichert-NOAA** [comment]
Yes, that sounds reasonable-- It's easy enough to set a build-time environment variable within Spack.

**Comment by jswhit** [comment]
@AlexanderRichert-NOAA PR #1470 adds this.  I have not tested it yet.

**Comment by AlexanderRichert-NOAA** [comment]
I tested building with Spack and I can confirm it disables the parallel support. Once this is in a release, I'm happy to update or help update the Spack recipe (which will need a little tweaking anyway, as it stills [requires matching of mpi enablement](https://github.com/spack/spack-packages/blob/14570534a39b7261bd6f2d4ddce750773e62f96e/repos/spack_repo/builtin/packages/py_netcdf4/package.py#L52) with netcdf-c and hdf5 even with the patching).
